devdoc/interlocked_linux_requirements.md

```diff
@@ -0,0 +1,345 @@
+MOCKABLE_FUNCTION(, int32_t, interlocked_xor, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_xor_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_xor_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_xor_8, volatile int8_t*, destination, int8_t, value);
+```
+
+## interlocked_add
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_add, volatile int32_t*, addend, int32_t, value);
+```
+**SRS_INTERLOCKED_LINUX_43_001: [** `interlocked_add` shall call `atomic_fetch_add` from `stdatomic.h`. **]**
+
+**SRS_INTERLOCKED_LINUX_43_002: [** `interlocked_add` shall return the result of the addition. **]**
+
+## interlocked_add_64
+
+```c
+MOCKABLE_FUNCTION(, int64_t, interlocked_add_64, volatile int64_t*, addend, int64_t, value);
+```
+**SRS_INTERLOCKED_LINUX_43_064: [** `interlocked_add_64` shall call `atomic_fetch_add` from `stdatomic.h`. **]**
+
+**SRS_INTERLOCKED_LINUX_43_065: [** `interlocked_add_64` shall return the result of the addition. **]**
+
+## interlocked_and
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_and, volatile int32_t*, destination, int32_t, value);
+```
+**SRS_INTERLOCKED_LINUX_43_003: [** `interlocked_and` shall call `atomic_fetch_and` from `stdatomic.h`. **]**
+
+**SRS_INTERLOCKED_LINUX_43_004: [** `interlocked_and` shall return the initial value of `*destination`. **]**
+
+## interlocked_and_16
+
+```c
+MOCKABLE_FUNCTION(, int16_t, interlocked_and_16, volatile int16_t*, destination, int16_t, value);
+```
+**SRS_INTERLOCKED_LINUX_43_005: [** `interlocked_and_16` shall call `atomic_fetch_and` from `stdatomic.h`. **]**
+
+**SRS_INTERLOCKED_LINUX_43_006: [** `interlocked_and_16` shall return the initial value of `*destination`. **]**
+ 
+ ## interlocked_and_64
+
+```c
+MOCKABLE_FUNCTION(, int64_t, interlocked_and_64, volatile int64_t*, destination, int64_t, value);
+```
+**SRS_INTERLOCKED_LINUX_43_007: [** `interlocked_and_64` shall call `atomic_fetch_and` from `stdatomic.h`. **]**
+
+**SRS_INTERLOCKED_LINUX_43_008: [** `interlocked_and_64` shall return the initial value of `*destination`. **]**
+
+## interlocked_and_8
+
+```c
+MOCKABLE_FUNCTION(, int8_t, interlocked_and_8, volatile int8_t*, destination, int8_t, value);
+```
+**SRS_INTERLOCKED_LINUX_43_009: [** `interlocked_and_8` shall call `atomic_fetch_and` from `stdatomic.h`. **]**
+
+**SRS_INTERLOCKED_LINUX_43_010: [** `interlocked_and_8` shall return the initial value of `*destination`. **]**
+
+## interlocked_compare_exchange
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_compare_exchange, volatile int32_t*, destination, int32_t, exchange, int32_t, comperand);
+```
+**SRS_INTERLOCKED_LINUX_43_011: [** `interlocked_compare_exchange` shall call `atomic_compare_exchange_strong` from `stdatomic.h`. **]**
+
+**SRS_INTERLOCKED_LINUX_43_012: [** `interlocked_compare_exchange` shall return the initial value of `*destination`. **]**

```

Comment:
>initial value of `*destination` [](start = 84, length = 31)

I also think these need to be explicit (especially because there is some logic to handle here)

If `atomic_compare_exchange_strong` returns `true` then `interlocked_compare_exchange` shall return `comperand`.

If `atomic_compare_exchange_strong` returns `false` then:

 - `interlocked_compare_exchange` shall call `atomic_load_explicit` to get the current value of `destination`.

 - If the loaded value is different than `comparand` then `interlocked_compare_exchange` shall return the loaded value.

Of course this leads to an issue we probably need to address here. What if `atomic_compare_exchange_strong` returns `false` but `atomic_load_explicit` returns `comperand`? You can't return `comperand` if the compare-exchange failed (because that breaks the contract for this interface that we return the value only if we did the exchange). So you either need to:

1. Change this function to return a flag like C11 atomics do (this can then easily be implemented either way), but then callers we convert from InterlockedCompareExchange need to have big code changes.
2. If the load after compare exchange is not consistent, then loop over and repeat this. Not quite as performant in all cases, but should be correct. This means you will need to add more requirements here of course

Reply:
If 'atomic_compare_exchange_strong' returns false, why do we have to check that the loaded value is different than 'comparand'? Why will it be the same? What should happen if it is the same?


---
In reply to: [431445821](https://github.com/Azure/azure-c-util/pull/44#discussion_r431445821) [](ancestors = 431445821)

Reply:
Discussed offline:

1. Need to check because we need to to atomic_compare_exchange and atomic_fetch as two separate instructions, so the value can change in between
2. If we don't check and we return `comparand` but atomic_compare_exchange returned false, then we broke our contract (we should only return the same value as `comparand` if the exchange happened.
3. If it is the same, we need to retry (do the atomic_compare_exchange and possibly the atomic_fetch again)

---
In reply to: [431491999](https://github.com/Azure/azure-c-util/pull/44#discussion_r431491999) [](ancestors = 431491999,431445821)